src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml

```diff
@@ -871,25 +891,41 @@ struct
         let (s1, x1), (s2, x2) = (red v1, red v2) in
         match (x1, x2) with
         | `Var x1, `Var x2 ->
-            (* s1 x1 - s2 x2 = 0
+            if Fp.equal s1 s2 then (
+              if not (Fp.equal s1 Fp.zero) then
+                Union_find.union (union_find sys x1) (union_find sys x2) )
+            else if (* s1 x1 - s2 x2 = 0
           *)
-            if not (Fp.equal s1 s2) then
+                    not (Fp.equal s1 s2) then
               add_generic_constraint ~l:x1 ~r:x2
                 [| s1; Fp.(negate s2); Fp.zero; Fp.zero; Fp.zero |]
                 sys
-              (* TODO: optimize by not adding generic costraint but rather permuting the vars *)
             else
               add_generic_constraint ~l:x1 ~r:x2
                 [| s1; Fp.(negate s2); Fp.zero; Fp.zero; Fp.zero |]
                 sys
-        | `Var x1, `Constant ->
-            add_generic_constraint ~l:x1
-              [| s1; Fp.zero; Fp.zero; Fp.zero; Fp.negate s2 |]
-              sys
-        | `Constant, `Var x2 ->
-            add_generic_constraint ~r:x2
-              [| Fp.zero; s2; Fp.zero; Fp.zero; Fp.negate s1 |]
-              sys
+        | `Var x1, `Constant -> (
+            (* s1 * x1 = s2
+               x1 = s2 / s1
+            *)
+            match Hashtbl.find sys.cached_constants Fp.(s2 / s1) with
```

Comment:
Need to be careful about division by zero here. We probably need to update `accumulate_sorted_terms` to filter out any terms whose coefficients are zero, or similar, otherwise we'll get a panic from rust here.